enhancements/cloud-integration/openstack/cloud-provider-config-upgrade.md

```diff
@@ -0,0 +1,441 @@
+They have configured a separate secret containing their cloud credentials as
+well as additional config directives. Specifically, `openshift-config/cloud-provider-config` contains:
+
+```ini
+[Global]
+secret-name = dick-jones-openstack-credentials
+secret-namespace = ed209-project
+
+...
+```
+
+Upon attempting to upgrade, they discover they cannot as the Cluster Config
+Operator is marked as non-upgradable. The status of this operator informs them
+that they must return `secret-name` and `secret-namespace` to their default
+values before they can upgrade.
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+
+N/A
+
+### Risks and Mitigations
+
+**Risk** Identical config options could work differently between the legacy and
+external cloud providers, or the conversion could be lossy.
+
+**Mitigation** Auditing and testing of the configuration options and
+combinations of same. At a minimum, we will explicitly test every legacy config
+option listed in the user documentation.
+
+**Risk** There may be additional configuration required for CCM compared to the
+legacy cloud provider in certain cloud configurations.
+
+**Mitigation** We currently test multiple cloud configurations. We should test
+an upgraded cluster on each of these cloud configurations.
+
+## Design Details
+
+### Configuration transformation for OpenStack CCM
+
+The configuration of OpenStack CCM is a superset of the legacy cloud provider
+configuration, with four exceptions. There are no features which are supported
+by the legacy cloud provider which are not supported by OpenStack CCM. Details
+of how we will handle the four removed configuration directives are provided
+below.
+
+<dl>
+<dt>secret-name</dt>
+<dt>secret-namespace</dt>
+<dt>kubeconfig-path</dt>
+<dd>
+  These are no longer relevant or supported in CCM. They define the cloud
+  credentials that are used by the legacy cloud provider. CCCMO will
+  automatically replace these with static config pointing to the cluster-managed
+  cloud credentials.
+  <br>
+  It is possible that the secret contains more information that the cloud
+  credentials populated by the installer. If so, the updated configuration will
+  no longer include these manual edits. Adding additional information to the
+  cloud credentials secret is not supported as we do not test or document it.
+  <br>
+  We estimate that the false positive risk of attempting to validate all
+  possible OpenStack authentication configurations is significantly higher than
+  the likelihood of any user actually having done this. We therefore judge that
+  it is safer not to attempt to handle this situation explicitly, and simply
+  constrain the user to modify their configuration to use the cluster cloud
+  credentials. If any user had put non-credential information in the credentials
+  secret, this would also be a prompt to them to fix that.
+</dd>
+<dt>node-volume-attach-limit</dt>
```

Comment:
According to docs, `node-volume-attach-limit` is [still available in the Cinder CSI configuration](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/cinder-csi-plugin/using-cinder-csi-plugin.md#block-storage).

`trust-device-path` [is gone](https://kubernetes-docsy-staging.netlify.app/docs/concepts/cluster-administration/cloud-providers/#block-storage) however.

Reply:
Done